src/build.ts

```diff
@@ -77,29 +77,79 @@ export async function build(nitro: Nitro) {
     : _build(nitro, rollupConfig);
 }
 
+const DYNAMIC_PARAM_RE = /(\*\*)(?!:)|((\*\*)?:[^/]+)/g;
+
 export async function writeTypes(nitro: Nitro) {
   const routeTypes: Record<
     string,
     Partial<Record<RouterMethod | "default", string[]>>
   > = {};
 
+  const fetchSignatures = [] as Array<[route: string, type: string]>;
+  const eventHandlerImports = new Set<string>();
+
   const typesDir = resolve(nitro.options.buildDir, "types");
 
   const middleware = [...nitro.scannedHandlers, ...nitro.options.handlers];
 
+  let i = 1;
   for (const mw of middleware) {
     if (typeof mw.handler !== "string" || !mw.route) {
       continue;
     }
-    const relativePath = relative(typesDir, mw.handler).replace(
-      /\.[a-z]+$/,
-      ""
-    );
+    const relativePath = relative(
+      typesDir,
+      resolveAlias(mw.handler, nitro.options.alias)
+    ).replace(/\.[a-z]+$/, "");
 
     if (!routeTypes[mw.route]) {
       routeTypes[mw.route] = {};
     }
 
+    const eventHandlerType = `EventHandler${i++}`;
+    eventHandlerImports.add(
+      `type ${eventHandlerType} = typeof import('${relativePath}').default`
+    );
+
+    // const isOptionsOptional
+    const isMethodOptional = !mw.method || mw.method.toUpperCase() === "GET";
+    const excludedMethods = middleware
+      .filter(
+        (other) => other.method && other.route === mw.route && other !== mw
+      )
+      .flatMap((m) =>
+        [m.method.toUpperCase(), m.method.toLowerCase()].map((m) => `'${m}'`)
+      )
+      .join(" | ");
+
+    const defaultMethod = excludedMethods
+      ? `Exclude<DefaultMethod, ${excludedMethods}>`
+      : "DefaultMethod";
+
+    const methodType = mw.method
+      ? [mw.method.toUpperCase(), mw.method.toLowerCase()]
+          .filter((m) => m !== "patch")
+          .map((m) => `'${m}'`)
+          .join(" | ")
+      : defaultMethod;
+
+    // TODO: 1. Fine-tune matching algorithm?
+    // TODO: merge return types
+    // TODO: 2. infer returns when we provide typed input
```

Comment:
Partial Inference is only beginning to be worked on in 5.2 (https://github.com/microsoft/TypeScript/pull/26349). 

Currently, identity functions are used to get this behaviour, providing sth like a "type closure": 
- https://medium.hexlabs.io/partial-type-inference-f669ae24c44e
- https://medium.com/@nandin-borjigin/partial-type-argument-inference-in-typescript-and-workarounds-for-it-d7c772788b2e

Reply:
Yes, this may only be possible with the new object-syntax event handlers: https://github.com/unjs/h3/pull/485